wolk/service/file_management/FileTransferSession.h

```diff
@@ -51,110 +54,125 @@ class FileTransferSession
     /**
      * Default constructor for the FileTransferSession in case of a regular file transfer.
      *
+     * @param deviceKey The device key for which this session is going on.
      * @param message The message that initiated an upload.
      * @param callback The callback that the session should use to announce status and error changes.
      * @param commandBuffer The command buffer which the session will use to announce status.
      */
-    FileTransferSession(const FileUploadInitiateMessage& message,
-                        std::function<void(FileUploadStatus, FileUploadError)> callback, CommandBuffer& commandBuffer);
+    FileTransferSession(std::string deviceKey, const FileUploadInitiateMessage& message,
+                        std::function<void(FileTransferStatus, FileTransferError)> callback,
+                        CommandBuffer& commandBuffer);
 
     /**
      * Default constructor for the FileTransferSession in case of a url download transfer.
      *
+     * @param deviceKey The device key for which this session is going on.
      * @param message The message that initiated an upload.
      * @param callback The callback that the session should use to announce status and error changes.
      * @param commandBuffer The command buffer which the session will use to announce status.
      * @param fileDownloader The file downloader that will actually execute the file download.
      */
-    FileTransferSession(const FileUrlDownloadInitMessage& message,
-                        std::function<void(FileUploadStatus, FileUploadError)> callback, CommandBuffer& commandBuffer,
-                        std::shared_ptr<FileDownloader> fileDownloader);
+    FileTransferSession(std::string deviceKey, const FileUrlDownloadInitMessage& message,
+                        std::function<void(FileTransferStatus, FileTransferError)> callback,
+                        CommandBuffer& commandBuffer, std::shared_ptr<FileDownloader> fileDownloader);
+
+    /**
+     * Default virtual destructor.
+     */
+    virtual ~FileTransferSession() = default;
 
     /**
      * Default getter for the information if the session is a platform transfer session.
      *
      * @return Is this session a platform transfer session.
      */
-    bool isPlatformTransfer() const;
+    virtual bool isPlatformTransfer() const;
```

Comment:
Couldn't we implement this with polymorphism?

Reply:
Well, the tinniest issue is that when a `file_binary_response` is received, we'd need to cast the object into a `PlatformFileTransferSession`, to pass it the message. That's why the two sessions kinda can't have a same interface.